### ISSUE TITLE
support common table expression (CTE)

### ISSUE THREAD

**Issue by zz-jason** [pingcap/tidb#17472]
## Description

**Is your feature request related to a problem? Please describe:**

Related issues:
- https://github.com/pingcap/tidb/issues/6824 request to support recursive common table expression
- https://github.com/pingcap/tidb/issues/17916 request to support common table expression to run TPC-DS and other analytic workloads
- https://github.com/pingcap/parser/issues/289 request to support common table expression to be compatible with MySQL 8.0

**Describe the feature you'd like:**

Support the common table expression feature like MySQL [13.2.15 WITH (Common Table Expressions)](https://dev.mysql.com/doc/refman/8.0/en/with.html):

> A common table expression (CTE) is a named temporary result set that exists within the scope of a single statement and that can be referred to later within that statement, possibly multiple times. The following discussion describes how to write statements that use CTEs.

There are two kinds of CTEs:

- non-recursive CTE
- recursive CTE

A typical query example of non-recursive CTE:

```sql
WITH
  cte1 AS (SELECT a, b FROM table1),
  cte2 AS (SELECT c, d FROM table2)
SELECT b, d FROM cte1 JOIN cte2
WHERE cte1.a = cte2.c;
```

A typical query example of recursive CTE:

```sql
WITH RECURSIVE cte (n) AS
(
  SELECT 1
  UNION ALL
  SELECT n + 1 FROM cte WHERE n < 5
)
SELECT * FROM cte;
```

**Comment by pingyu** [comment]
Let me try...

**Comment by pingyu** [comment]
Find these:
- [Implementing Common Table Expressions for MariaDB](https://seim-conf.org/media/materials/2017/proceedings/SEIM-2017_Full_Papers.pdf#page=13)

- [MDEV-8789 Implement non-recursive common table expressions](https://github.com/MariaDB/server/pull/134)
- [MDEV-9864 - Implement recursive common table expressions](https://github.com/MariaDB/server/pull/182)

**Comment by tisonkun** [comment]
Here is the link to weekly report of CTE development: https://docs.google.com/document/d/1b-nL0c2akLL03ZhXXcQT6tj_tqQMmfAE8oKSfcaaDqE

I'm wondering whether we can just post weekly update on this thread as below on finalize:

## 2021/04/22 (10%)

### Summary

- [x] parser/pull/1207 Parser part gets merged into the master branch. @wjhuang2016
- [x] tidb/pull/24170: support new variable cte_max_recursion_depth @xiongjiwei
- [ ] (IN REVIEW) tidb/pull/24147: CTE design proposal @guo-shaoge
- [ ] (IN REVIEW) tidb/pull/24193: CTEStorage implementation @guo-shaoge
- [ ] (WIP) Demo for CTE, finish the logical plan and physical plan. @wjhuang2016

### Next Week plan

- Finish the demo, make the basic CTE statement work.
- Finish the CTE support for randgen test

**Comment by morgo** [comment]
@wjhuang2016 should we close this issue now since CTEs are in master?

**Comment by tisonkun** [comment]
> @wjhuang2016 should we close this issue now since CTEs are in master?

A feature implemented without design doc #24808 merged 😂

**Comment by wjhuang2016** [comment]
The design doc is merged. So I close it now.
